lib/bank_account.ex

```diff
@@ -8,31 +10,78 @@ defmodule BankAccount do
   """
   @opaque account :: pid
 
+  @impl true
+  def init(:ok) do
+    {:ok, %{balance: 0, status: :account_open}}
+  end
+
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok)
+  end
+
+  @impl true
+  # balance inquiry when status: :account_closed
+  def handle_call(:balance, _from, %{status: :account_closed} = state) do
+    {:reply, {:error, :account_closed}, state}
+  end
+
+  @impl true
+  # balance inquiry
+  def handle_call(:balance, _from, state) do
+    {:reply, Map.get(state, :balance), state}
+  end
+
+  @impl true
+  # close bank/update status
+  def handle_call({:status, :account_closed}, _from, state) do
+    {:reply, :ok, Map.put(state, :status, :account_closed)}
+  end
+
+  @impl true
+  # balance update when status: account_closed
+  def handle_call({:balance, _amount}, _from, %{status: :account_closed} = state) do
+    {:reply, {:error, :account_closed}, state}
+  end
+
+  @impl true
+  # balance update
+  def handle_call({:balance, amount}, _from, state) do
+    balance = Map.get(state, :balance)
+    {:reply, :ok, Map.put(state, :balance, balance + amount)}
+  end
+
   @doc """
   Open the bank. Makes the account available.
   """
   @spec open_bank() :: account
+  # TODO: handle errors?
+  # @spec open_bank() :: {:ok, account} | {:error, term()}
   def open_bank() do
+    {:ok, account} = BankAccount.start_link()
+    account
   end
 
   @doc """
   Close the bank. Makes the account unavailable.
   """
   # @spec close_bank(account) :: none
   def close_bank(account) do
+    GenServer.call(account, {:status, :account_closed})
```

Comment:
It is idiomatic in Elixir GenServers to have the "public" functions like this up at the top of the file, with the `start_link` function following them, then the `init` call, then the `handle_call`/`handle_cast` functions, and finally any internal functions that the `handle_call`/`handle_cast` use. This way a person jumping into the code sees the public stuff first, and doesn't have to dig through the internals to find it.